poetry/repositories/installed_repository.py

```diff
@@ -99,6 +101,127 @@ def is_vcs_package(cls, package: Union[Path, Package], env: Env) -> bool:
+                )
+                if paths:
+                    is_editable_package = False
+                    for src in paths:
+                        if cls.is_vcs_package(src, env):
+                            (
+                                source_type,
+                                source_url,
+                                source_reference,
+                            ) = cls.get_package_vcs_properties_from_path(src)
+                            break
+
+                        if not (
+                            is_editable_package or env.is_path_relative_to_lib(src)
+                        ):
+                            is_editable_package = True
+                    else:
+                        # TODO: handle multiple source directories?
+                        if is_editable_package:
+                            source_type = "directory"
+                            source_url = paths.pop().as_posix()
+        else:
+            if cls.is_vcs_package(path, env):
+                (
+                    source_type,
+                    source_url,
+                    source_reference,
+                ) = cls.get_package_vcs_properties_from_path(
+                    env.path / "src" / canonicalize_name(distribution.metadata["name"])
+                )
+            else:
+                # If not, it's a path dependency
+                source_type = "directory"
+                source_url = str(path.parent)
+
+        package = Package(
+            distribution.metadata["name"],
+            distribution.metadata["version"],
+            source_type=source_type,
+            source_url=source_url,
+            source_reference=source_reference,
+            source_resolved_reference=source_resolved_reference,
+        )
+        package.description = distribution.metadata.get("summary", "")
+
+        return package
+
+    @classmethod
+    def create_package_from_pep610(cls, distribution: "Distribution") -> Package:
+        path = Path(str(distribution._path))
+        source_type = None
+        source_url = None
+        source_reference = None
+        source_resolved_reference = None
+        develop = False
+
+        url_reference = json.loads(
+            path.joinpath("direct_url.json").read_text(encoding="utf-8")
+        )
+        if "archive_info" in url_reference:
+            # File or URL distribution
+            if url_reference["url"].startswith("file:"):
+                # File distribution
+                source_type = "file"
+                source_url = Path(
+                    unquote(urlparse(url_reference["url"]).path)
+                ).as_posix()
+            else:
+                # URL distribution
+                source_type = "url"
+                source_url = url_reference["url"]
+        elif "dir_info" in url_reference:
+            # Directory distribution
+            source_type = "directory"
+            source_url = Path(url_reference["url"]).as_posix()
```

Comment:
We should use `poetry.core.utils.url_to_path` falling back to treating as `path`.